staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime.go

```diff
@@ -161,3 +176,134 @@ func RecoverFromPanic(err *error) {
 			callers)
 	}
 }
+
+func newDedupingErrorHandler(cacheSize, errorDepth int, delta time.Duration) *dedupingErrorHandler {
+	d := &dedupingErrorHandler{
+		cache: lru.New(cacheSize),
+		count: make(map[errKey]errVal),
+
+		errorDepth: errorDepth,
+		delta:      delta,
+		clock:      clock.RealClock{},
+	}
+
+	d.cache.OnEvicted = func(key lru.Key, _ interface{}) {
+		// remove the associated entry in the count map when this key is evicted from the LRU cache
+		delete(d.count, key.(errKey))
+	}
+
+	return d
+}
+
+// dedupingErrorHandler provides a go routine safe error handler via handleErr.
+// It tracks errors via the caller stack, the error type and the error message (err.Error() value).
+// An error is considered unique based on these properties (see errKey).
+// To prevent from using an infinite amount of memory, it uses a LRU cache to purge old error values.
+type dedupingErrorHandler struct {
+	mutex sync.Mutex
+
+	// cache tracks (stack + type + message) and cleans up old entries in count as they roll off the cache
+	cache *lru.Cache
+	// count tracks (stack + type + message) -> (count + logged)
+	// since rudimentaryErrorBackoff rate limits HandleError to 1000 errors/second,
+	// this counter will effectively never overflow (nothing bad happens even if it does)
+	count map[errKey]errVal
+
+	// errorDepth is how many frames to skip from handleErr
+	errorDepth int
+
+	// delta is the minimum difference between the current time and
+	// errVal.logged required for us to log the associated error again
+	delta time.Duration
+
+	// clock allows us to control time in unit tests
+	// TODO add unit tests
+	clock clock.Clock
+}
+
+// errKey tracks uniqueness based on the caller's stack and the type/message of the error
+// it is stored in both the count map and the LRU cache
+// it is removed from the count map when it gets evicted from the LRU cache
+// it is comparable via ==
+type errKey struct {
+	stack   string
+	errType reflect.Type
+	// message is the err.Error() value
+	message string
+}
+
+// errVal tracks how many times we have seen an error, and the last time we logged it
+type errVal struct {
+	count  uint64
+	logged time.Time
+}
+
+// handleErr logs the given error if it is considered new or "not recently logged"
+// currently it logs errors whenever:
+// 1. the associated errKey does not exist in d.count
+// 2. the associated counter is a power of two
+// 3. the associated logged time's delta from the current time is greater than d.delta
+func (d *dedupingErrorHandler) handleErr(err error) {
+	d.mutex.Lock()
```

Comment:
Ok, so this is why it works.  That's a really big hammer, but it is an error case.  I won't dismiss it out of hand.